exotations/task_maps/task_map/src/LookAt.cpp

```diff
@@ -0,0 +1,101 @@
+ *
+ * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
+ * AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
+ * IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
+ * ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT OWNER OR CONTRIBUTORS BE
+ * LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR
+ * CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
+ * SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
+ * INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
+ * CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
+ * ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+ * POSSIBILITY OF SUCH DAMAGE.
+ *
+ */
+
+#include "task_map/LookAt.h"
+
+REGISTER_TASKMAP_TYPE("LookAt", exotica::LookAt);
+
+namespace exotica
+{
+LookAt::LookAt() = default;
+LookAt::~LookAt() = default;
+
+Eigen::Vector3d LookAt::getLookAtTarget()
+{
+    return Eigen::Map<Eigen::Vector3d>(Kinematics[0].Phi(2).p.data);
+}
+
+void LookAt::update(Eigen::VectorXdRefConst x, Eigen::VectorXdRef phi)
+{
+    if (phi.rows() != 3) throw_named("Wrong size of phi!");
+
+    // Get EffPoint and LookAtTarget
+    Eigen::Vector3d c = Eigen::Map<Eigen::Vector3d>(Kinematics[0].Phi(0).p.data);  // EffPoint | Eff frame
+    Eigen::Vector3d p = Eigen::Map<Eigen::Vector3d>(Kinematics[0].Phi(1).p.data);  // LookAtTarget | Eff frame
+
+    // Compute orthogonal orthogonal projection a onto line e->c
+    double c_squared_norm = c.squaredNorm();
+    double p_dot_c = p.dot(c);
+    double alpha = p_dot_c / c_squared_norm;
+    Eigen::Vector3d a = alpha * c;
+
+    // Set phi
+    phi = a - p;
+}
+
+void LookAt::update(Eigen::VectorXdRefConst x, Eigen::VectorXdRef phi, Eigen::MatrixXdRef J)
+{
+    if (phi.rows() != 3) throw_named("Wrong size of phi!");
+    if (J.rows() != 3 || J.cols() != Kinematics[0].J(0).data.cols()) throw_named("Wrong size of J! " << Kinematics[0].J(0).data.cols());
+
+    // Get EffPoint and LookAtTarget
+    Eigen::Vector3d c = Eigen::Map<Eigen::Vector3d>(Kinematics[0].Phi(0).p.data);  // EffPoint | Eff frame
+    Eigen::Vector3d p = Eigen::Map<Eigen::Vector3d>(Kinematics[0].Phi(1).p.data);  // LookAtTarget | Eff frame
+
+    // Compute orthogonal orthogonal projection a onto line e->c
+    double c_squared_norm = c.squaredNorm();
+    double p_dot_c = p.dot(c);
+    double alpha = p_dot_c / c_squared_norm;
+    Eigen::Vector3d a = alpha * c;
+
+    // Set phi
+    phi = a - p;
+
+    // Compute J
+    for (int j = 0; j < J.cols(); j++)
+    {
+        Eigen::Vector3d pd = Kinematics[0].J[1].data.topRows<3>().col(j);
+        double alphad = c.dot(pd) / c_squared_norm;
+        J.col(j) = alphad * c - pd;
+    }
+}
+
+void LookAt::Instantiate(LookAtInitializer &init)
+{
+    if (Frames.size() < 3) throw_named("Three frames are required!");
+}
+
+int LookAt::taskSpaceDim()
```

Comment:
`int ... const`

Reply:
Please could you clarify. Do you mean line 97 should be as follows?
```
int ... const LookAt::taskSpaceDim()
```

Reply:
`int LookAt::taskSpaceDim() const` - also in header. It is a const method as it does not change private members of the class when calling

Reply:
I am getting the following error when I compile:
```
Errors     << task_map:make /home/chris/Projects/kuka_ws/logs/task_map/build.make.218.log                                                                                                                                             
In file included from /home/chris/Projects/kuka_ws/src/exotica/exotations/task_maps/task_map/src/LookAt.cpp:33:0:
/home/chris/Projects/kuka_ws/src/exotica/exotations/task_maps/task_map/include/task_map/LookAt.h:71:9: error: ‘int exotica::LookAt::taskSpaceDim() const’ marked ‘override’, but does not override
     int taskSpaceDim() const override;
         ^
In file included from /home/chris/Projects/kuka_ws/src/exotica/exotica/include/exotica/TaskMap.h:36:0,
                 from /home/chris/Projects/kuka_ws/src/exotica/exotations/task_maps/task_map/include/task_map/LookAt.h:36,
                 from /home/chris/Projects/kuka_ws/src/exotica/exotations/task_maps/task_map/src/LookAt.cpp:33:
/home/chris/Projects/kuka_ws/src/exotica/exotations/task_maps/task_map/src/LookAt.cpp: In lambda function:
/home/chris/Projects/kuka_ws/src/exotica/exotica/include/exotica/Factory.h:52:124: error: invalid new-expression of abstract class type ‘exotica::LookAt’
     static exotica::Registrar<BASE> EX_UNIQ(object_registrar_, __LINE__)("exotica/" TYPE, []() -> BASE* { return new DERIV(); }, #BASE); \
                                                                                                                            ^
/home/chris/Projects/kuka_ws/src/exotica/exotica/include/exotica/TaskMap.h:48:44: note: in expansion of macro ‘EXOTICA_REGISTER’
 #define REGISTER_TASKMAP_TYPE(TYPE, DERIV) EXOTICA_REGISTER(exotica::TaskMap, TYPE, DERIV)
                                            ^
/home/chris/Projects/kuka_ws/src/exotica/exotations/task_maps/task_map/src/LookAt.cpp:35:1: note: in expansion of macro ‘REGISTER_TASKMAP_TYPE’
 REGISTER_TASKMAP_TYPE("LookAt", exotica::LookAt);
 ^
In file included from /home/chris/Projects/kuka_ws/src/exotica/exotations/task_maps/task_map/src/LookAt.cpp:33:0:
/home/chris/Projects/kuka_ws/src/exotica/exotations/task_maps/task_map/include/task_map/LookAt.h:41:7: note:   because the following virtual functions are pure within ‘exotica::LookAt’:
 class LookAt : public TaskMap, public Instantiable<LookAtInitializer>
       ^
In file included from /home/chris/Projects/kuka_ws/src/exotica/exotations/task_maps/task_map/include/task_map/LookAt.h:36:0,
                 from /home/chris/Projects/kuka_ws/src/exotica/exotations/task_maps/task_map/src/LookAt.cpp:33:
/home/chris/Projects/kuka_ws/src/exotica/exotica/include/exotica/TaskMap.h:63:17: note: 	virtual int exotica::TaskMap::taskSpaceDim()
     virtual int taskSpaceDim() = 0;
                 ^
In file included from /opt/ros/kinetic/include/class_loader/class_loader_core.hpp:46:0,
                 from /opt/ros/kinetic/include/class_loader/./class_loader.hpp:43,
                 from /opt/ros/kinetic/include/class_loader/class_loader.h:35,
                 from /opt/ros/kinetic/include/pluginlib/./class_list_macros.hpp:40,
                 from /opt/ros/kinetic/include/pluginlib/class_list_macros.h:35,
                 from /home/chris/Projects/kuka_ws/src/exotica/exotica/include/exotica/Factory.h:42,
                 from /home/chris/Projects/kuka_ws/src/exotica/exotica/include/exotica/TaskMap.h:36,
                 from /home/chris/Projects/kuka_ws/src/exotica/exotations/task_maps/task_map/include/task_map/LookAt.h:36,
                 from /home/chris/Projects/kuka_ws/src/exotica/exotations/task_maps/task_map/src/LookAt.cpp:33:
/opt/ros/kinetic/include/class_loader/meta_object.hpp: In instantiation of ‘B* class_loader::class_loader_private::MetaObject<C, B>::create() const [with C = exotica::LookAt; B = exotica::TaskMap]’:
/home/chris/Projects/kuka_ws/src/exotica/exotations/task_maps/task_map/src/LookAt.cpp:117:1:   required from here
/opt/ros/kinetic/include/class_loader/meta_object.hpp:198:16: error: invalid new-expression of abstract class type ‘exotica::LookAt’
     return new C;
                ^
make[2]: *** [CMakeFiles/task_map.dir/src/LookAt.cpp.o] Error 1
make[1]: *** [CMakeFiles/task_map.dir/all] Error 2
make: *** [all] Error 2
cd /home/chris/Projects/kuka_ws/build/task_map; catkin build --get-env task_map | catkin env -si  /usr/bin/make --jobserver-fds=6,7 -j; cd -
```
I am guessing deeper changes need to be made to exotica?